flink-core/src/main/java/org/apache/flink/util/AbstractCloseableRegistry.java

```diff
@@ -61,13 +75,14 @@ public final void registerClosable(C closeable) throws IOException {
 		}
 
 		synchronized (getSynchronizationLock()) {
-			if (closed) {
-				IOUtils.closeQuietly(closeable);
-				throw new IOException("Cannot register Closeable, registry is already closed. Closing argument.");
+			if (!closed) {
+				doRegister(closeable, closeableToRef);
+				return;
 			}
-
-			doRegister(closeable, closeableToRef);
 		}
+
+		IOUtils.closeQuietly(closeable);
```

Comment:
Should this also be in synchronized block?

Reply:
I think it should not. For some distributed filesystems a call to `close()` could block and would keep the registry blocked. We only need to synchronize the access to the internal map and the closed flag.

Reply:
In general, the idea of the change is to move every call that could block or aquire other locks, with a potential to deadlock outside of the synchronized blocks, so that this class will never block for other threads.

Reply:
I see

Reply:
I'm wondering, should it actually be the responsibility of the registry to close the `closeable` if it's already closed or should it be the responsibility of however wants to register that closable?

Reply:
I have this done on purpose. For the time a ``Closeable`` is registered, it is managed by the registry but I do not want to stop somebody from removing a ``Closeable`` again and reclaim responsibility over the resource. I was, however, considering a convenience method that does the remove-and-close in one call.

Reply:
And for the second part, I think it should be closed automatically, because you the caller decides to give the responsibility to the registry and tie it to the registry's status. So the registry should take care that the close status is propagated to new incoming objects.

Reply:
Makes sense! 👌